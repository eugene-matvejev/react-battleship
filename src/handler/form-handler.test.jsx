import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FormHandler from './form-handler';

configure({ adapter: new Adapter() });

describe('<FormHandler/>', () => {
    const ctrlProps = {
        updateCTRL: {
            label: '{{update}}',
        },
        submitCTRL: {
            label: '{{submit}}',
        },
        cancelCTRL: {
            label: '{{cancel}}',
        },
    };

    const props = {
        config: [
            {
                title: '{{section 0}}',
                items: [
                    {
                        c: (props) => <span {...props} />,
                    }
                ]
            }
        ],
        ...ctrlProps,
    };

    const data = {};

    const e = {
        preventDefault: () => 0,
        stopPropagation: () => 0,
        target: {
            value: '{{value}}',
            getAttribute: () => 0,
        },
    };

    describe('render', () => {
        it('with default/required props', () => {
            const c = shallow(<FormHandler {...props} />);

            expect(c).toMatchSnapshot();
        });

        describe('with optional props', () => {
            [
                ['title', '{{title}}'],
                ['data-cy', '{{data-cy}}'],
                ['data', data],
            ].forEach(([prop, v]) => {
                it(`[::${prop}] as "${v}"`, () => {
                    const c = shallow(<FormHandler {...props} {...{ [prop]: v }} />);

                    expect(c).toMatchSnapshot();
                });
            });
        });
    });

    describe('lifecycle events', () => {
        describe('::componentDidMount', () => {
            it('should invoke external callback [::onMount]', () => {
                const spy = jest.fn();

                const c = shallow(<FormHandler {...props} onMount={spy} />);

                expect(spy).toBeCalledWith(c.instance().props, c.state(), c.instance().onSuccess, c.instance().onError);
            });
        });
    });

    describe('internal callbacks', () => {
        describe('::onSuccess', () => {
            it('should set state fields [::config, ::data] from payload', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');

                shallow(<FormHandler {...props} />).instance().onSuccess({ data, config: props.config });

                expect(spy).toBeCalledWith({ data, config: props.config });
            });

            it('should invoke external callback [::onSuccess]', () => {
                const spy = jest.fn();

                const c = shallow(<FormHandler {...props} onSuccess={spy} />);
                c.instance().onSuccess({ data, config: props.config });

                expect(spy).toBeCalledWith(c.instance().props, c.state());
            });
        });

        describe('::onError', () => {
            it('should set state fields [::config, ::data] from payload', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');

                shallow(<FormHandler {...props} />).instance().onError({ data, config: props.config });

                expect(spy).toBeCalledWith({ data, config: props.config });
            });

            it('should invoke external callback [::onError]', () => {
                const spy = jest.fn();

                const c = shallow(<FormHandler {...props} onError={spy} />);
                c.instance().onError({ data, config: props.config });

                expect(spy).toBeCalledWith(c.instance().props, c.state());
            });
        });

        describe('::onSubmit', () => {
            it('should invoke external callback [::onSubmit] from click on [data-cy="form-action-submit"]', () => {
                const spy = jest.fn();
                const c = shallow(<FormHandler {...props} onSubmit={spy} />);

                c.find('Button[data-cy="form-action-submit"]').simulate('click', e);

                expect(spy).toBeCalledWith(c.instance().props, c.state(), c.instance().onSuccess, c.instance().onError);
            });

            it('should invoke external callback [::onSubmit] from click on [data-cy="form-action-update"]', () => {
                const spy = jest.fn();
                const c = shallow(<FormHandler {...props} onSubmit={spy} data={data} />);

                c.find('Button[data-cy="form-action-update"]').simulate('click', e);

                expect(spy).toBeCalledWith(c.instance().props, c.state(), c.instance().onSuccess, c.instance().onError);
            });

            it('should invoke external callback [::validate] from click on [data-cy="form-action-submit"]', () => {
                const spy = jest.fn();
                const c = shallow(<FormHandler {...props} validate={spy} />);

                c.find('Button[data-cy="form-action-submit"]').simulate('click', e);

                expect(spy).toBeCalledWith(c.state('config'));
            });

            it('when external callback [::validate] undefined, state field [::isValid] should be set to true', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');

                const c = shallow(<FormHandler {...props}  />);

                c.find('Button[data-cy="form-action-submit"]').simulate('click', e);

                expect(spy).toBeCalledWith({ isValid: true });
            });

            it('when external callback [::validate] return false, state field [::isValid] should be set to false, and [::config] field should get new reference', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');

                const c = shallow(<FormHandler {...props} validate={() => false} />);

                c.find('Button[data-cy="form-action-submit"]').simulate('click', e);

                expect(spy).toBeCalledWith({ config: c.state('config'), isValid: false });
            });
        });

        describe('::onCancel', () => {
            it('should invoke external callback [::onCancel] from click on [data-cy="form-action-cancel"]', () => {
                const spy = jest.fn();
                const c = shallow(<FormHandler {...props} onCancel={spy} />);

                c.find('Button[data-cy="form-action-cancel"]').simulate('click', e);

                expect(spy).toBeCalledWith(c.instance().props, c.state(), c.instance().onSuccess, c.instance().onError);
            });
        });

        /** functionality relay on data-section attribute */
        describe('::onCollapse', () => {
            it('should mutate state field [::config] and toggle [::isCollapsed] field for relevant section, from click on <Accordion/>', () => {
                const c = shallow(<FormHandler {...props} />);

                c.find('[data-section][data-cy="section-0"]').simulate('collapse', e);

                expect(c.state('config')).toMatchSnapshot();
            });
        });

        /** functionality relay on data-section and data-field attributes */
        describe('::onChange', () => {
            it('should invoke external callback [::validate] with relevant payload', () => {
                const spy = jest.fn();
                const c = shallow(<FormHandler {...props} validate={spy} />);

                c.find('[data-section][data-field][data-cy="section-0-input-0"]').simulate('change', e);

                expect(spy).toBeCalledWith(c.state('config'), [[0, 0]]);
            });

            it('should mutate state field [::config] and set [::value] for relevant field from payload', () => {
                const config = [
                    {
                        items: [
                            {
                                c: () => <span />,
                            },
                            {
                                c: () => <span />,
                            },
                        ],
                    },
                ];

                const c = shallow(<FormHandler {...props} config={config} />);

                c.find('[data-section][data-field][data-cy="section-0-input-0"]').simulate('change', e);

                expect(c.state('config')).toMatchSnapshot();
            });

            it('should set state field [::isValid] to FALSE, if there is least one error', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');
                const config = [
                    {
                        items: [
                            {
                                c: () => <span />,
                            },
                            {
                                c: () => <span />,
                                errors: ['error'],
                            },
                        ],
                    },
                ];

                shallow(<FormHandler {...props} config={config} />)
                    .find('[data-section][data-field][data-cy="section-0-input-0"]')
                    .simulate('change', e);

                expect(spy).toBeCalledWith({ config, isValid: false });
            });

            it('should set state field [::isValid] to TRUE, if there are NO errors', () => {
                const spy = spyOn(FormHandler.prototype, 'setState');
                const config = [
                    {
                        items: [
                            {
                                c: () => <span />,
                            },
                            {
                                c: () => <span />,
                            },
                        ],
                    },
                ];

                shallow(<FormHandler {...props} config={config} />)
                    .find('[data-section][data-field][data-cy="section-0-input-0"]')
                    .simulate('change', e);

                expect(spy).toBeCalledWith({ config, isValid: true });
            });
        });
    });
});
