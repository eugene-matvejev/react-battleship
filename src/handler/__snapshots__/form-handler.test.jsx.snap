// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormHandler/> internal callbacks ::onChange should mutate state field [::config] and set [::value] for relevant field from payload 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "c": [Function],
        "value": "{{value}}",
      },
      Object {
        "c": [Function],
      },
    ],
  },
]
`;

exports[`<FormHandler/> internal callbacks ::onCollapse should mutate state field [::config] and toggle [::isCollapsed] field for relevant section, from click on <Accordion/> 1`] = `
Array [
  Object {
    "isCollapsed": true,
    "items": Array [
      Object {
        "c": [Function],
      },
    ],
    "title": "{{section 0}}",
  },
]
`;

exports[`<FormHandler/> render with default/required props 1`] = `
<form
  className="form "
>
  <Accordion
    className=""
    data-cy="section-0"
    data-section={0}
    key="0"
    onCollapse={[Function]}
    title="{{section 0}}"
  >
    <c
      data-cy="section-0-input-0"
      data-field={0}
      data-section={0}
      key="0"
      onChange={[Function]}
    />
  </Accordion>
  <div
    className="form_controls"
  >
    <Button
      className=""
      data-cy="form-action-submit"
      disabled={true}
      label="{{submit}}"
      onClick={[Function]}
    />
    <Button
      className=""
      data-cy="form-action-cancel"
      label="{{cancel}}"
      onClick={[Function]}
    />
  </div>
</form>
`;

exports[`<FormHandler/> render with optional props [::data] as "[object Object]" 1`] = `
<form
  className="form "
>
  <Accordion
    className=""
    data-cy="section-0"
    data-section={0}
    key="0"
    onCollapse={[Function]}
    title="{{section 0}}"
  >
    <c
      data-cy="section-0-input-0"
      data-field={0}
      data-section={0}
      key="0"
      onChange={[Function]}
    />
  </Accordion>
  <div
    className="form_controls"
  >
    <Button
      className=""
      data-cy="form-action-update"
      disabled={true}
      label="{{update}}"
      onClick={[Function]}
    />
    <Button
      className=""
      data-cy="form-action-cancel"
      label="{{cancel}}"
      onClick={[Function]}
    />
  </div>
</form>
`;

exports[`<FormHandler/> render with optional props [::data-cy] as "{{data-cy}}" 1`] = `
<form
  className="form "
>
  <Accordion
    className=""
    data-cy="{{data-cy}}section-0"
    data-section={0}
    key="0"
    onCollapse={[Function]}
    title="{{section 0}}"
  >
    <c
      data-cy="{{data-cy}}section-0-input-0"
      data-field={0}
      data-section={0}
      key="0"
      onChange={[Function]}
    />
  </Accordion>
  <div
    className="form_controls"
  >
    <Button
      className=""
      data-cy="{{data-cy}}form-action-submit"
      disabled={true}
      label="{{submit}}"
      onClick={[Function]}
    />
    <Button
      className=""
      data-cy="{{data-cy}}form-action-cancel"
      label="{{cancel}}"
      onClick={[Function]}
    />
  </div>
</form>
`;

exports[`<FormHandler/> render with optional props [::title] as "{{title}}" 1`] = `
<form
  className="form "
>
  <h1
    className="form_title"
    data-cy="form-title"
  >
    {{title}}
  </h1>
  <Accordion
    className=""
    data-cy="section-0"
    data-section={0}
    key="0"
    onCollapse={[Function]}
    title="{{section 0}}"
  >
    <c
      data-cy="section-0-input-0"
      data-field={0}
      data-section={0}
      key="0"
      onChange={[Function]}
    />
  </Accordion>
  <div
    className="form_controls"
  >
    <Button
      className=""
      data-cy="form-action-submit"
      disabled={true}
      label="{{submit}}"
      onClick={[Function]}
    />
    <Button
      className=""
      data-cy="form-action-cancel"
      label="{{cancel}}"
      onClick={[Function]}
    />
  </div>
</form>
`;
